File: frontend/lib/interfaces.ts
import { TablerIconsProps } from "@tabler/icons-react";
import { ObjectId } from "mongodb";
import { FC } from "react";

export interface INavigationItem {
    label: string; // Label of the item
    to: string; // Where it goes
    icon: FC<TablerIconsProps>; // Icon
    desc: string; // A short description
    perm: number; // Permission number: 2=public, 3=admin
}

export interface IEvaluation {
    tournament_name: string;
    round_name: string;
    is_prelim: boolean;
    is_improvement: boolean;
    decision: number;
    comparison: number;
    citation: number;
    coverage: number;
    bias: number;
    weight: number;
    date: Object;
    div_name: string;
}

export interface IParadigm {
    nationality: string;
    gender: string;
    age: string;
    university: string;
}

export interface IJudge {
    _id: ObjectId;
    name: string;
    email: string;
    evaluations: IEvaluation[];
    paradigm: string;
    options: IParadigm;
}

File: backend/src/models/interfaces.rs
//models/interfaces.rs
use serde::{Deserialize, Serialize};
use mongodb::bson::DateTime;
use bson::oid::ObjectId;

// types of data
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct IEvaluation {
    pub tournament_name: String,
    pub round_name: String,
    pub is_prelim: bool,
    pub is_improvement: bool,
    pub div_name: String,
    
    pub decision: f32,
    pub comparison: f32,
    pub citation: f32,
    pub coverage: f32,
    pub bias: f32,
    pub weight: f32,
    pub date: DateTime
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct IParadigm {
    pub nationality: String,
    pub gender: String,
    pub age: String,
    pub university: String
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct IJudge {
    pub _id: ObjectId,
    pub name: String,
    pub email: String,
    pub evaluations: Vec<IEvaluation>,
    pub paradigm: Option<String>,
    pub options: IParadigm
}

#[derive(Debug, Deserialize, Serialize)]
pub struct IUser {
    pub _id: ObjectId,
    pub email: String,
    pub name: String,
    pub password: String,  // This should store a hashed version of the password, never plain text
    pub permission_level: i32,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct IFrontendUser {
    pub email: String,
    pub name: String,
    pub permission_level: i32,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct IInviteCode {
    pub _id: ObjectId,
    pub code: String,
    pub email: String,
    pub permission_level: i32,
    pub name: String
}
